include/cyclic_reduction.hxx

```diff
@@ -100,30 +103,33 @@ public:
   ///                where N is set in the constructor or setup
   /// @param[in] b   Diagonal values. Should have size [nsys][N]
   /// @param[in] c   Right diagonal. Should have size [nsys][N]
-  void setCoefs(int nsys, T **a, T **b, T **c) {
+  void setCoefs(int nsys, Matrix<T> &a, Matrix<T> &b, Matrix<T> &c) {
+    TRACE("CyclicReduce::setCoefs");
+    
     // Make sure correct memory arrays allocated
     allocMemory(nprocs, nsys, N);
 
     // Fill coefficient array
     for(int j=0;j<Nsys;j++)
       for(int i=0;i<N;i++) {
-        coefs[j][4*i] = a[j][i];
-        coefs[j][4*i + 1] = b[j][i];
-        coefs[j][4*i + 2] = c[j][i];
+        coefs(j, 4*i) = a(j, i);
+        coefs(j, 4*i + 1) = b(j, i);
+        coefs(j, 4*i + 2) = c(j, i);
         // 4*i + 3 will contain RHS
       }
   }
 
   /// Solve a single triadiagonal system
   /// 
-  void solve(T rhs[], T x[]) {
+  void solve(Matrix<T> &rhs, Matrix<T> &x) {
```

Comment:
I think this should be Array, rather than Matrix. This function solves a single system (Ax=b), so `x` and `rhs` here should be Arrays. On line 128 the address of `rhs` and `x` were used, to convert T* to T** . Is there a way to convert an Array to a single-column (or single-row) Matrix?

Reply:
There's not currently a way to convert an Array to a `Matrix`, but it should be easy enough to add a constructor to Matrix that does this. In general there isn't a unique way to do this, but we could always make the choice that either the matrix is 1xN or Nx1.

I've been thinking about the possibility of making `Matrix::operator[]` return an `Array` slice of the data (either by creating a new Array and copying in values, or making `Matrix::data` actually be a vector of `Array`). WIth this I think it would make sense to get a `1xN` shape Matrix when constructing from an Array of size N.

This interface to `solve` isn't actually used anywhere at the moment (at least not in any of the framework or the tests) so we could remove it and revisit if/when this is needed/the Matrix<-->Array conversion is sorted?

Reply:
I should note there's a similar argument for setcoefs.

Reply:
I've added interfaces to `setCoefs` and `solve` that take `Array` instead of `Matrix`. These then copy into a `Matrix` and pass to the main routine before copying back into the `Array` (where needed).